include/openPMD/IO/ADIOS/ADIOS2IOHandler.hpp

```diff
@@ -21,64 +21,595 @@
 #include <string>
+#include <unordered_map>
+#include <utility> // pair
+#include <vector>
+
+
+#if openPMD_HAVE_ADIOS2
+#include <adios2.h>
+#include "openPMD/IO/ADIOS/ADIOS2Auxiliary.hpp"
+#endif
+
+#if openPMD_HAVE_MPI
+#include <mpi.h>
+#endif
 
 
 namespace openPMD
 {
 #if openPMD_HAVE_ADIOS2
+
 class ADIOS2IOHandler;
 
+namespace detail
+{
+    template < typename, typename > struct DatasetHelper;
+    struct DatasetReader;
+    struct AttributeReader;
+    struct AttributeWriter;
+    template < typename > struct AttributeTypes;
+    struct DatasetOpener;
+    template < typename > struct DatasetTypes;
+    struct WriteDataset;
+    struct BufferedActions;
+    struct BufferedPut;
+    struct BufferedGet;
+    struct BufferedAttributeRead;
+} // namespace detail
+
+
 class ADIOS2IOHandlerImpl
+: public AbstractIOHandlerImplCommon< ADIOS2FilePosition >
 {
+    template < typename, typename > friend struct detail::DatasetHelper;
+    friend struct detail::DatasetReader;
+    friend struct detail::AttributeReader;
+    friend struct detail::AttributeWriter;
+    template < typename > friend struct detail::AttributeTypes;
+    friend struct detail::DatasetOpener;
+    template < typename > friend struct detail::DatasetTypes;
+    friend struct detail::WriteDataset;
+    friend struct detail::BufferedActions;
+    friend struct detail::BufferedAttributeRead;
+
+    static constexpr bool ADIOS2_DEBUG_MODE = false;
+
+
 public:
-    ADIOS2IOHandlerImpl(AbstractIOHandler*);
-    virtual ~ADIOS2IOHandlerImpl();
-
-    virtual std::future< void > flush();
-
-    using ArgumentMap = std::map< std::string, ParameterArgument >;
-    virtual void createFile(Writable*, ArgumentMap const&);
-    virtual void createPath(Writable*, ArgumentMap const&);
-    virtual void createDataset(Writable*, ArgumentMap const&);
-    virtual void extendDataset(Writable*, ArgumentMap const&);
-    virtual void openFile(Writable*, ArgumentMap const&);
-    virtual void openPath(Writable*, ArgumentMap const&);
-    virtual void openDataset(Writable*, ArgumentMap &);
-    virtual void deleteFile(Writable*, ArgumentMap const&);
-    virtual void deletePath(Writable*, ArgumentMap const&);
-    virtual void deleteDataset(Writable*, ArgumentMap const&);
-    virtual void deleteAttribute(Writable*, ArgumentMap const&);
-    virtual void writeDataset(Writable*, ArgumentMap const&);
-    virtual void writeAttribute(Writable*, ArgumentMap const&);
-    virtual void readDataset(Writable*, ArgumentMap &);
-    virtual void readAttribute(Writable*, ArgumentMap &);
-    virtual void listPaths(Writable*, ArgumentMap &);
-    virtual void listDatasets(Writable*, ArgumentMap &);
-    virtual void listAttributes(Writable*, ArgumentMap &);
-
-    AbstractIOHandler* m_handler;
-};  //ADIOS2IOHandlerImpl
-#else
-class ADIOS2IOHandlerImpl
-{ };
-#endif
+    static_assert(
+        sizeof( bool ) == 1,
+        "ADIOS2 backend needs a platform with boolean size equals one byte." );
```

Comment:
this makes me shiver. even if a `bool` is one size, a [`std::vector< bool >`](https://en.cppreference.com/w/cpp/container/vector_bool) can be anything. Are we safe? :)

Reply:
There is no `VEC_BOOL` [datatype](https://github.com/openPMD/openPMD-api/blob/b4a904b6f1d0ce38294ab499066acfdd53fdc00a/include/openPMD/Datatype.hpp#L38) in openPMD, so this issue does not arise.
I added this check just to be sure that a single bool can be stored in an unsigned char, but I think the check is a bit pointless and can be removed.

Reply:
Sounds good, thanks for double-checking!